### PR TITLE
Project generator

### PR DESCRIPTION
**Pull request by latrokles**
## Overview

- implements a project generator for TaskCat. It currently only supports
  the Quick Start project type, but can be extended further by creating
  the template structure under `templates/${project_type}`.

- the generator is rather simple, it uses a set of predefined jinja templates
  that it traverses, renders, and saves into the project's directory.

- the configuration object is a simple named tuple that defines instance
  scoped values that are used to render the template.

- the existing configuration is pretty basic, I used the arguments we
  currently use in `examples/sample-taskcat-project`, but maybe each
  project type should provide its own configuration definition.

## Testing/Steps taken to ensure quality

- while I've included a unit test for the happy path, I've also included a
  simple runner as a sort of functional/integ test until we establish
  better infrastructure for integration tests. This can be removed once
  we integrate the generator with the cli option parser.

### Notes

The ProjectConfiguration namedtuple can be extracted into a separate file as QuickStartProjectConfig or QuickStartProjectValues. It may be the case that other project types will use a different set of values that can be passed in to their respective templates.

## Testing Instructions

1. unit test: python setup.py test
2. hand-rolled integ test
```
$ python taskcat/project_generator_runner.py /tmp/new_project
$ tree /tmp/new_project
/tmp/new_project
├── README.md
├── ci
│   ├── debug-input.json
│   ├── taskcat-autobucket-json.yml
│   ├── taskcat-autobucket.yml
│   └── taskcat.yml
├── scripts
│   └── scripts_userdata.sh
└── templates
    ├── debug-yaml.template
    └── debug.template

3 directories, 8 files

```
addresses #209

### CODE REVIEW

**Comment by latrokles** [comment]
So tests are failing in travis due to import issues, but they pass running `python setup.py test`.

Are we testing differently on travis and locally? is some documentation on how to go about this?

File: taskcat/project_generator.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        if t.endswith(TEMPLATE_FILE_EXTENSION)]
+
+
+ProjectConfiguration = namedtuple(
+    'ProjectConfiguration',
+    'owner_email, project_name, project_type, supported_regions'
+)
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
```
Comment (jaymccon) [inline taskcat/project_generator.py]: might be better not to hardcode this and let the consumer decide on log level

File: taskcat/project_generator.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            project_path = self._full_destination_path(directory)
+            self._make_project_directory(project_path)
+
+            template_filepaths = template_paths(directory, files)
+            self._generate_project_files(template_filepaths, project_path)
+
+    def _traverse_templates(self):
+        return self.filesystem.traverse_templates(self.config.project_type)
+
+    def _full_destination_path(self, path):
```
[the remark this answers is not in the document]
Comment (latrokles) [reply taskcat/project_generator.py]: oh this is a good point, I guess technically it's not overriding it as the imported `path` is still used outside of this method w/out issues.

In the method it's used as a method scoped variable inside, but could totally lead to a bug if we intended to used the imported `path` from os inside that method.

This is - however - a recipe for this disaster if we ever need to use the imported `path` (from `os`) inside the method. I'll rename it.

File: taskcat/project_generator.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        )
+        path = path.replace(templates_root, '')
+        return full_path(self.destination, path)
+
+    def _make_project_directory(self, project_directory):
+        try:
+            logger.info('creating {}'.format(project_directory))
+            self.filesystem.create_project_directory(project_directory)
+        except FileExistsError as e:
+            logging.warn("{} - skipping...".format(e))
```
[the remark this answers is not in the document]
Comment (latrokles) [reply taskcat/project_generator.py]: good to know, thanks.

File: taskcat/project_generator.py (hunk 1 above)
Comment (latrokles) [reply taskcat/project_generator.py]: that's fair.

File: taskcat/project_generator.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+    def _generate_project_files(self, template_filepaths, project_directory):
```
[the remark this answers is not in the document]
Comment (latrokles) [reply taskcat/project_generator.py]: oh yeah, left over from a revision that was using it to generate destination paths. Half expected my editor to catch that as unused var.

Removing it.

**Comment by agamzn** [comment]
@latrokles  - Thanks. Formal PR commentary forthcoming.

**Comment by latrokles** [comment]
@agamzn 

> The templates themselves shouldn't be included here. They're examples, but they're also used in integration testing of taskcat via travis-ci. They don't reflect what a template skeleton will look like for the quickstart project type. I'd rather see no data here, than what we have now. There's a pretty big confusion risk if someone is poking around the repository. As an alternate, I'm fine with a README that indicates we need to add more content here to flush it out.

I'm confused as to what you are referring to here. Is it:
a) the `.template` files 
b) everything under `project_templates`

I'm not sure why is it a problem to include data here. If we settle on what a barebones, starter Quick Start project is, we can template that and bundle in, and generate such project from that.

>The same can be said for the taskcat parameters. Automatically generating these will be addressed in #86. Ditto with the README preference.

>I'd also like to see logic around a single taskcat config file. How do we apply conditional keys (marketplace: true, for example., autobucket vs pre-defined bucket, as another)

Not quite sure what this is about.

All the points above seem to stem from miscommunication. A few months ago I sat down with @avattathil and prototyped this into a single python file, from every discussion we've had so far it was my understanding that we wanted that into taskcat. This is the more readable, maintainable implementation of the code in the gist I circulated back then.

> The FileSystemService class contains instance functions, but I don't see a class instantiation anywhere.

If you look at `project_generator_runner.py` you will see how this class is used and passed into the `ProjectGenerator` in order to perform filesystem operations. The idea is the piece of taskcat that determines this feature will  used will instantiate the `ProjectGenerator` and pass an instance of `FilesystemService`. This was created so that we could unit test `ProjectGenerator` without having to touch the filesystem (we know this is brittle).

**Comment by agamzn** [comment]
@latrokles 

> I'm confused as to what you are referring to here. Is it:
> a) the `.template` files
> b) everything under `project_templates`
> 

I'm referring to everything under `taskcat/taskcat/project_templates/quickstart/templates`. 

> I'm not sure why is it a problem to include data here. If we settle on what a barebones, starter Quick Start project is, we can template that and bundle in, and generate such project from that.
> 

 As a team, we haven't had that discussion yet. We've agreed that there should be a template, but not what that template should be.  I'd rather see *no* data, then data we haven't come to a consensus on. 

In my mind, a quickstart-specific skeleton will contain boilerplate parameters (`QSS3KeyPrefix`, `QSS3BucketName`), an example of a nested template, an example of a call to S3, so forth. Everything under `examples/sample-taskcat-project` is used to demonstrate what taskcat can do, and also doubles as a functional test for us in Travis. They're not necessarily a demonstration of what a boilerplate quickstart should look like. 

> > I'd also like to see logic around a single taskcat config file. How do we apply conditional keys (marketplace: true, for example., autobucket vs pre-defined bucket, as another)
> 
> Not quite sure what this is about.
>

There are multiple examples of what a taskcat config file *could* look like here:
- https://github.com/aws-quickstart/taskcat/tree/5869fd9f2859b1ffb20116069b7c86889416c31e/taskcat/project_templates/quickstart/ci

We need a way to conditionalize these into a single config. Examples include..
- are use using an auto-generated bucket, or using the `s3bucket` key to point to a static bucket
- is the quickstart leveraging govcloud? 
- is the quickstart using a marketplace AMI?
and others.

How do we do that with the current PR? 

> All the points above seem to stem from miscommunication. A few months ago I sat down with @avattathil and prototyped this into a single python file, from every discussion we've had so far it was my understanding that we wanted that into taskcat. This is the more readable, maintainable implementation of the code in the gist I circulated back then.
> 

That's a fair point. I think it's prudent to have the conversation before we merge in a PR, though. I'd rather spend a few days making sure everyone understands what the intent of a PR is, rather than assume, merge,  and have issues later. :)

> > The FileSystemService class contains instance functions, but I don't see a class instantiation anywhere.
> 
> If you look at `project_generator_runner.py` you will see how this class is used and passed into the `ProjectGenerator` in order to perform filesystem operations. The idea is the piece of taskcat that determines this feature will used will instantiate the `ProjectGenerator` and pass an instance of `FilesystemService`. This was created so that we could unit test `ProjectGenerator` without having to touch the filesystem (we know this is brittle).

After digging, I understand this. Thanks! 👍

**Comment by latrokles** [comment]
> @latrokles
> 
> Removing the templates discussion, because we can add in what we actually need later

I get now what you meant (templates is an overloaded term in here), we can remove them now or later, the only thing w/ removing everything under that directory is that git will remove the empty dir altogether (we could put an empty `.keepme` to keep the dir).

As a whole I'd defer to you all on what goes where as you would know better, I just wanted to provide an initial skeleton to iterate on.

> > > I'd also like to see logic around a single taskcat config file. How do we apply conditional keys (marketplace: true, for example., autobucket vs pre-defined bucket, as another)
> > 
> > 
> > Not quite sure what this is about.
> 
> There are multiple examples of what a taskcat config file _could_ look like here:
> 
> * https://github.com/aws-quickstart/taskcat/tree/5869fd9f2859b1ffb20116069b7c86889416c31e/taskcat/project_templates/quickstart/ci
> 
> We need a way to conditionalize these into a single config. Examples include..
> 
> * are use using an auto-generated bucket, or using the `s3bucket` key to point to a static bucket
> * is the quickstart leveraging govcloud?
> * is the quickstart using a marketplace AMI?
>   and others.
> 
> How do we do that with the current PR?
> 
We don't, but reading the above a few things come to mind.

1) make a more featured config object for quick starts that covers these cases.
2) put all the conditional logic on the template. For example:
```
{% if config.some_parameter is True %}
... render this chunk of config
{% endif %}
```
and so on.
I prefer the generational conditional to live on the templates themselves.

> > > The FileSystemService class contains instance functions, but I don't see a class instantiation anywhere.
> > 
> > 
> > If you look at `project_generator_runner.py` you will see how this class is used and passed into the `ProjectGenerator` in order to perform filesystem operations. The idea is the piece of taskcat that determines this feature will used will instantiate the `ProjectGenerator` and pass an instance of `FilesystemService`. This was created so that we could unit test `ProjectGenerator` without having to touch the filesystem (we know this is brittle).
> 
> After digging, I understand this. Thanks! 👍
np.

**Comment by jaymccon** [comment]
> I get now what you meant (templates is an overloaded term in here), we can remove them now or later, the only thing w/ removing everything under that directory is that git will remove the empty dir altogether (we could put an empty `.keepme` to keep the dir).
> 
> As a whole I'd defer to you all on what goes where as you would know better, I just wanted to provide an initial skeleton to iterate on.

My vote is to keep what's there and then start working to improve it, with the aim of providing a batteries included way to initialize a new project. Any other use cases (like `taskcat`ifying an existing project) should be a separate project type, and imho are a lesser priority.

**Comment by agamzn** [comment]
> > I get now what you meant (templates is an overloaded term in here), we can remove them now or later, the only thing w/ removing everything under that directory is that git will remove the empty dir altogether (we could put an empty `.keepme` to keep the dir).
> > As a whole I'd defer to you all on what goes where as you would know better, I just wanted to provide an initial skeleton to iterate on.
> 
> My vote is to keep what's there and then start working to improve it, with the aim of providing a batteries included way to initialize a new project. Any other use cases (like `taskcat`ifying an existing project) should be a separate project type, and imho are a lesser priority.

:+1: to Jay's commentary above.


> > There are multiple examples of what a taskcat config file _could_ look like here:
> > 
> > * https://github.com/aws-quickstart/taskcat/tree/5869fd9f2859b1ffb20116069b7c86889416c31e/taskcat/project_templates/quickstart/ci
> > 
> > We need a way to conditionalize these into a single config. Examples include..
> > 
> > * are use using an auto-generated bucket, or using the `s3bucket` key to point to a static bucket
> > * is the quickstart leveraging govcloud?
> > * is the quickstart using a marketplace AMI?
> >   and others.
> > 
> > How do we do that with the current PR?
> 
> We don't, but reading the above a few things come to mind.
> 
> 1. make a more featured config object for quick starts that covers these cases.
> 2. put all the conditional logic on the template. For example:
> 
> ```
> {% if config.some_parameter %}
> ... render this chunk of config
> {% endif %}
> ```
> and so on.
> I prefer the generational conditional to live on the templates themselves.
> 
> > > > The FileSystemService class contains instance functions, but I don't see a class instantiation anywhere.
> > > 
> > > 
> > > If you look at `project_generator_runner.py` you will see how this class is used and passed into the `ProjectGenerator` in order to perform filesystem operations. The idea is the piece of taskcat that determines this feature will used will instantiate the `ProjectGenerator` and pass an instance of `FilesystemService`. This was created so that we could unit test `ProjectGenerator` without having to touch the filesystem (we know this is brittle).
> > 
> > 
> > After digging, I understand this. Thanks! 👍
> > np.

@latrokles  - I concur, let's get this PR merged, and I'll submit another one that consolidates the conditionals.

**Pull request merged**